File: ios/Nativebrik/Nativebrik/data/component.swift
//
//  component.swift
//  Nativebrik
//
//  Created by Ryosuke Suzuki on 2024/03/06.
//

import Foundation

protocol ComponentRepository2 {
    func fetchComponent(experimentId: String, id: String) async -> Result<UIBlock, NativebrikError>
}

class ComponentRepositoryImpl: ComponentRepository2 {
    private let config: Config
    init(config: Config) {
        self.config = config
    }

    func fetchComponent(experimentId: String, id: String) async -> Result<UIBlock, NativebrikError> {
        guard let url = URL(string: config.cdnUrl + "/projects/" + config.projectId + "/experiments/components/" + experimentId + "/" + id) else {
            return Result.failure(NativebrikError.irregular("Failed to create URL object"))
        }
        
        let data = await getData(url: url)
        switch data {
        case .success(let data):
            let decoder = JSONDecoder()
            guard let result = try? decoder.decode(UIBlockJSON.self, from: data) else {
                return Result.failure(NativebrikError.failedToDecode)
            }
            return Result.success(result)
            
        case .failure(let error):
            return Result.failure(error)
        }
    }
}

File: android/nativebrik/src/main/java/com/nativebrik/sdk/data/component.kt
package com.nativebrik.sdk.data

import com.nativebrik.sdk.Config
import com.nativebrik.sdk.schema.UIBlock
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

internal interface ComponentRepository {
    suspend fun fetchComponent(experimentId: String, id: String): Result<UIBlock>
}

internal class ComponentRepositoryImpl(private val config: Config, private val cache: Cache): ComponentRepository {
    override suspend fun fetchComponent(experimentId: String, id: String): Result<UIBlock> {
        return withContext(Dispatchers.IO) {
            val url = config.endpoint.cdn + "/projects/" + config.projectId + "/experiments/components/" + experimentId + "/" + id
            val response: String = getRequestWithCache(url, cache).getOrElse {
                return@withContext Result.failure(it)
            }
            val json = Json.decodeFromString<JsonElement>(response)
            val configs = UIBlock.decode(json) ?: return@withContext Result.failure(FailedToDecodeException())
            return@withContext Result.success(configs)
        }
    }
}
